File: src/CastleClashdb/Services/HeroService.cs
using CastleClashdb.Models.CastleClash;
using CastleClashdb.Repositories;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace CastleClashdb.Services
{
    public class HeroService : IDisposable
    {
        private IGenericRepository repo;
        public HeroService(IGenericRepository repo)
        {
            this.repo = repo;
        }
        //get heroes
        public IEnumerable<Hero> GetAllHeroes()
        {
            return repo.Query<Hero>();
        }
        // bool version
        public bool GetHeroById(int id, out Hero item)
        {
            item = repo.Query<Hero>().FirstOrDefault(m => m.Id == id);
            return item != null;
        }
        public Hero GetHeroById(int id)
        {
            return repo.Query<Hero>().FirstOrDefault(m => m.Id == id);
        }

        //create
        public Hero SaveHero(Hero item)
        {
            return repo.Add<Hero>(item);
                
        }

        //update
        public void UpdateHero(int id, Hero item)
        {
            item.Id = id;
            repo.Update<Hero>(item);
        }

        //Delete 
        public void DeleteHero(int id)
        {
            var item = repo.Query<Hero>().FirstOrDefault(m => m.Id == id);
            repo.Delete<Hero>(item);
        }
            
        //Cleanup
        public void Dispose()
        {
            repo.Dispose();
        }
    }
}

File: src/CastleClashdb/wwwroot/js/services/hero.service.ts
namespace CastleClashdb.Services {

    export class ResourceService {

        protected updateResourceAction: ng.resource.IActionHash = {
            'update': { method: 'PUT' }
        };
    }
    export class HeroService extends ResourceService {
        private heroResource: IUpdatableResourceClass<Models.Hero>;

        static $inject = [
            '$resource',
            '$http'
        ];

        constructor(
            $resource: ng.resource.IResourceService,
            private $http: ng.IHttpService
        ) {
            super();
            this.heroResource = $resource<Models.Hero, IHeroResourceClass>('api/hero/:id', null, this.updateResourceAction);
        }
        //get all heroes
        public getAllHeroes(): Models.Hero[] {

            return this.heroResource.query();
        }

        //get
        public getHero(id: number): Models.Hero {
            return this.heroResource.get({ id: id });
        }

        // save a new hero
        public saveHero(hero: Models.Hero): ng.IPromise<Models.Hero> {
            return this.$http.post<Models.Hero>('api/hero/', hero)
                .then((response) => {
                    console.log(response.data);
                    return response.data;
                });
        }

        //update Hero
        public updateHero(hero: Models.Hero): void {
            this.heroResource.update({ id: hero.id }, hero);
        }

        //delete Hero
        public deleteHero(id: number) {
            this.heroResource.delete({ id: id });
        }


    }
    interface IHeroResourceClass extends IUpdatableResourceClass<Models.Hero> {

    }

    interface IUpdatableResourceClass<T> extends ng.resource.IResourceClass<T> {

        update(): T;
        update(params: Object): T;
        update(success: Function, error?: Function): T;
        update(params: Object, sucess: Function, error?: Function): T;
        update(params: Object, data: Object, success?: Function, error?: Function): T;
    }
}